### PR TITLE
Fix nullification of has_many :through association with `query_constraints

### PR DESCRIPTION
**Pull request by nvasilevski**
Given a `has_many :through` association with `query_constraints`:

```ruby
BlogPost.has_many :posts_tags
BlogPost.has_many :tags,
  through: :posts_tags,
  query_constraints: [:blog_id, :post_id]
```

It is possible to nullify the association like `blog_post.tags = []`
which results in deletion of records from the `posts_tags` joining table.

### Implementation details

The test we added fails on main with `NoMethodError: undefined method to_sym for ["blog_id", "id"]:Array` on the line 
https://github.com/rails/rails/blob/d53dc868869f973c7ee0bbcdaacfc0239f261773/activerecord/lib/active_record/associations/through_association.rb#L66

However this failure is misleading as we shouldn't be reaching this line at all, at least for the test we built.
In our case the issue lies a few lines above, at:
https://github.com/rails/rails/blob/d53dc868869f973c7ee0bbcdaacfc0239f261773/activerecord/lib/active_record/associations/through_association.rb#L62

Since the introduction of `query_constraints` in associations `association_primary_key` can be an array while `primary_key` may not necessarily be an array in case if the model uses virtual `query_constraints` definition. So essentially we need to compare `association_primary_key` with `query_constraints_list` if it is not empty and only then compare it with a primary key. To avoid branching I've introduced a new internal concept - `composite_query_constraints_list` this is an always non-empty array of query constraints that either equals  to manually configured `query_constraints` or to the `primary_key` wrapped in an array. This concept will be useful in the future in internal codeplaces that are ready to work with an array of columns to avoid unnecessary branching. 

After comparing primary keys as Arrays the logic follows into the branch and queries the records by the association name which has already been fixed in https://github.com/rails/rails/pull/47692

We still need to address the `else` branch but for this we will need to extend our `Sharded` fixtures to have an association with a custom `source_type` or `query_constraints:` configuration different from the parent's `query_constraints`

### CODE REVIEW

**Comment by eileencodes** [comment]
@nvasilevski can you take a look at the test failures?

File: activerecord/lib/active_record/persistence.rb [hunk 1]
```diff
@@ -499,6 +499,16 @@ def query_constraints_list # :nodoc:
         end
       end
 
+      # Always returns an array of column names to be used in querying.
+      # Regardless whether the source is the primary key or the query_constraints configuration.
+      # Supposed to be used internally in Rails in places where the primary key value can always be treated as an array.
+      # This simpifies the code as it avoids branches `on query_constraints_list.nil?`
+      # Long-term this method can be removed and behavior added to the query_constraints_list itself.
+      # However until then we need to keep `query_constraints_list` nillable for backwards compatibility.
```
Comment (eileencodes) [inline activerecord/lib/active_record/persistence.rb:507]: This documentation needs improvement, it's a bit more wordy than we need here. Can you replace it with something like

```
Returns an array of column names to be used in queries. The source of column 
names is derived from +query_constraints_list+ or +primary_key+. This method 
is for internal use when the primary key is to be treated as an array.
```

**Review by eileencodes — CHANGES_REQUESTED** [review]

**Comment by nvasilevski** [comment]
>can you take a look at the test failures?

Should be fixed now, I shouldn't have wrapped `association_primary_key` into an array for the whole flow. At this moment we should only wrap for the comparision

File: activerecord/lib/active_record/persistence.rb (hunk 1 above)
Comment (nvasilevski) [reply activerecord/lib/active_record/persistence.rb]: Changed

**Review by eileencodes — APPROVED** [review]

**Pull request merged**